src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -1,25 +1,102 @@
 package com.es.phoneshop.model.product;
 
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Currency;
 import java.util.List;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.stream.Collectors;
 
 public class ArrayListProductDao implements ProductDao {
-    @Override
-    public Product getProduct(Long id) {
-        throw new RuntimeException("Not implemented");
+  private long maxId;
+  private ReadWriteLock lock;
+  private List<Product> products;
+
+  public ArrayListProductDao() {
+    this.products = new ArrayList<>();
+    this.lock = new ReentrantReadWriteLock();
+    saveSampleProducts();
+  }
+
+  @Override
+  public Product getProduct(Long id) throws ProductNotFoundException {
+    Product result;
+    lock.readLock().lock();
+    try {
+      result = products.stream()
+              .filter(product -> product.getId().equals(id))
+              .findAny()
+              .orElseThrow(ProductNotFoundException::new);
+    } finally {
+      lock.readLock().unlock();
     }
+    return result;
+  }
 
-    @Override
-    public List<Product> findProducts() {
-        throw new RuntimeException("Not implemented");
+  @Override
+  public List<Product> findProducts() {
+    List<Product> productList;
+    lock.readLock().lock();
+    try {
+      productList = products.stream()
+              .filter(product -> product.getPrice() != null)
+              .filter(product -> product.getStock() > 0)
+              .collect(Collectors.toList());
+    } finally {
+      lock.readLock().unlock();
     }
+    return productList;
+  }
 
-    @Override
-    public void save(Product product) {
-        throw new RuntimeException("Not implemented");
+  @Override
+  public void save(Product product) {
+    Long productId = product.getId();
+    lock.writeLock().lock();
+    try {
+      if (productId == null) {
+        product.setId(maxId++);
+        products.add(product);
+      } else {
+        Product productToUpdate = getProduct(productId);
+        productToUpdate.setCode(product.getCode());
+        productToUpdate.setCurrency(product.getCurrency());
+        productToUpdate.setPrice(product.getPrice());
+        productToUpdate.setStock(product.getStock());
+        productToUpdate.setDescription(product.getDescription());
+        productToUpdate.setImageUrl(product.getImageUrl());
```

Comment:
Can we move this code to a separate method like populateExistingProduct() for better readability?